### PR TITLE
HTTP Client improvements, DNS Client rework, RFC 6335

### PR DESCRIPTION
**Pull request by AndreasAakesson**
For some reason the the callback delegate the HTTP Client sometimes ended up as nullptr. Maybe due to new delegate implementation. Seems to be fixed by using `spec::dynamic`. If anyone can explain why, please do :D

TCP and UDP are now using dynamic (ephemeral) ports in the range described by IANA in [RFC 6335](https://tools.ietf.org/html/rfc6335). In other words, the outgoing ports are now in range of `49152 - 65534`.

Improved the DNS Client by adding timeout on resolve requests (5s default). It now also reuses one UDP socket. Also added transaction ID matching on incoming responses.

Also a fix for #1073 sneaked in before I noticed #1074 - please merge that one first (when sorted).

### CODE REVIEW

File: src/net/tcp/tcp.cpp [hunk 1]
```diff
@@ -131,7 +131,7 @@ seq_t TCP::generate_iss() {
 */
 port_t TCP::next_free_port() {
 
-  current_ephemeral_ = (current_ephemeral_ == 0) ? current_ephemeral_ + 1025 : current_ephemeral_ + 1;
+  current_ephemeral_ = (current_ephemeral_ == port_ranges::DYNAMIC_END) ? port_ranges::DYNAMIC_START : current_ephemeral_ + 1;
 
   // Avoid giving a port that is bound to a service.
   while(listeners_.find(current_ephemeral_) != listeners_.end())
```
Comment (RostakaGmfun) [inline src/net/tcp/tcp.cpp]: I think the below line (`current_ephemeral_++;`) contains a bug. The value of `current_ephemeral_` may overflow the `port_ranges::DYNAMIC_END` if the last port from the dynamic port range is used by the listener. Am I right?

File: src/net/dns/client.cpp [hunk 2]
```diff
@@ -17,35 +17,49 @@
...
 {
+  Timer::duration_t DNSClient::DEFAULT_RESOLVE_TIMEOUT{std::chrono::seconds(5)};
+
   void DNSClient::resolve(IP4::addr dns_server, const std::string& hostname, Stack::resolve_func<IP4> func)
   {
-    auto& sock = stack.udp().bind();
-
     // create DNS request
     DNS::Request request;
     auto*  data = new char[256];
```
Comment (RostakaGmfun) [inline src/net/dns/client.cpp]: Could we allocate the `data` on stack if the size is known at compile time?

File: api/net/dns/dns.hpp [hunk 3]
```diff
@@ -149,22 +149,33 @@ namespace net
...
       const std::string& getHostname() const
       {
         return this->hostname;
       }
+
       IP4::addr getFirstIP4() const
       {
-        if (answers.size())
-          return answers[0].getIP4();
+        auto it = answers.begin();
```
Comment (RostakaGmfun) [inline api/net/dns/dns.hpp]: Have you considered using range-based for loop?

File: src/net/dns/client.cpp (hunk 2 above)
Comment (fwsGonzo) [inline src/net/dns/client.cpp]: Seems like it to me

**Comment by AndreasAakesson** [comment]
Thanks for the review @RostakaGmfun, much appreciated!

File: src/net/dns/client.cpp (hunk 2 above)
Comment (AndreasAakesson) [inline src/net/dns/client.cpp]: Yeah didn't think about that one.

File: api/net/dns/dns.hpp (hunk 3 above)
Comment (AndreasAakesson) [inline api/net/dns/dns.hpp]: Guess I somehow overlooked it, good suggestion.

File: src/net/tcp/tcp.cpp (hunk 1 above)
Comment (AndreasAakesson) [inline src/net/tcp/tcp.cpp]: I think you're right here yes. But it shouldn't be a listener on one of the dynamic ports (we do not prevent this tho, and I'm not sure if we want to do this either?). I know the TCP port-handling is not bulletproof, and that it maybe should be reworked some. 
What makes it a bit troublesome is that there are two collections where a port could be taken; `listeners_` (listening bound ports) and `connections_` (outgoing connections).
Another bug is when all dynamic ports are taken/active, we start overwriting existing ones. The port-handling in UDP is much more simpler/safer.

**Pull request merged**